### PR TITLE
[torch][ao] Do not crash numerics debugger if the shape of the tensors do not match

### PR DESCRIPTION
**Pull request by dulinriley**
Summary: Occasionally we see the loss function to crash because the shapes of `a` and `b` tensors are different. This diff avoids crashing is such scenarios and lets the comparison work for other nodes where the shapes match.

Test Plan: - CI

Reviewed By: jerryzh168

### CODE REVIEW

**Comment by dulinriley** [comment]
This is a redo of https://github.com/pytorch/pytorch/pull/141257 but with a different author who has permissions

**Review by nitish2112 — APPROVED** [review]

**Comment by dulinriley** [comment]
@jerryzh168 can you stamp this? I need an authorized reviewer

**Review by jerryzh168 — APPROVED** [review]
what are those cases? can you add a test btw?

**Comment by dulinriley** [comment]
> what are those cases? can you add a test btw?

This wasn't my diff, I commandeered it from someone who didn't have permissions to land, it was previously accepted internally.

This was mostly to avoid a very large exception message relating to printing tensors instead of their shapes

**Comment by jerryzh168** [comment]
> > what are those cases? can you add a test btw?
> 
> This wasn't my diff, I commandeered it from someone who didn't have permissions to land, it was previously accepted internally.
> 
> This was mostly to avoid a very large exception message relating to printing tensors instead of their shapes

I see, then should there be a raise Excpetion under the if statement

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 1]
```diff
@@ -318,10 +318,28 @@ def compare_results(
...
+            if (
+                isinstance(a, torch.Tensor)
+                and isinstance(b, torch.Tensor)
+                and a.shape != b.shape
+            ):
+                log.warning(
+                    "Cannot compare tensors with different shapes: actual=%s vs ref=%s",
+                    a.shape,
+                    b.shape,
+                )
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/_numeric_debugger.py:331]: should we raise exception here instead of ignore the error

File: torch/ao/quantization/pt2e/_numeric_debugger.py [hunk 2]
```diff
@@ -318,10 +318,28 @@ def compare_results(
...
             results = [
                 QuantizationComparisonResult(actual=a, ref=b)
                 for a, b in zip(actual_stats, ref_stats)
+                # Only compare objects if they're both collections, or both single
+                # tensors of the same shape.
+                # TODO: use torch.fx.node.map_aggregate for the collection of
+                # comparison results instead.
+                if not isinstance(a, torch.Tensor)
+                or not isinstance(b, torch.Tensor)
+                or a.shape == b.shape
```
Comment (jerryzh168) [inline torch/ao/quantization/pt2e/_numeric_debugger.py:342]: also with this, some of the mismatches will be ignored, that's probably not what we want I think

File: torch/ao/quantization/pt2e/_numeric_debugger.py (hunk 1 above)
Comment (dulinriley) [reply torch/ao/quantization/pt2e/_numeric_debugger.py:331]: I don't think so, because this type of error would typically indicate the debug handle was placed on the wrong nodes to compare before and after. Debug handles should always produce tensors of the same shape.

But debug handle errors should not prevent reporting of other accuracy errors.

File: torch/ao/quantization/pt2e/_numeric_debugger.py (hunk 2 above)
Comment (dulinriley) [reply torch/ao/quantization/pt2e/_numeric_debugger.py:342]: We want to ignore mismatches if they were comparing this of different types, or when the tensors have different shapes, because it's a problem with the debug handles at that point
